onnxruntime/core/providers/xnnpack/nn/softmax.cc

```diff
@@ -0,0 +1,195 @@
+          &p);
+    } else {
+      ORT_ENFORCE(0, "error kernel type input, expected uint8|float");
+    }
+    ORT_ENFORCE(xstatus == xnn_status_success, "xnn_create_softmax_nc_f32 failed. Status:", xstatus);
+    op0_.reset(p);
+  }
+
+  Status Compute(OpKernelContext* ctx) const override;
+
+ private:
+  Status ComputeImpl(const Tensor& input, Tensor& output, size_t axis,
+                     concurrency::ThreadPool* thread_pool) const;
+
+
+  Status ComputeImplOpset13(const Tensor& input, Tensor& output, size_t axis,
+                            concurrency::ThreadPool* thread_pool, OpKernelContext* ctx) const;
+
+  int axis_;
+  int opset_;
+  bool log_softmax_;
+  int32_t kernel_dtype = 0;
+  XnnpackOperator op0_ = nullptr;
+  enum InputTensors : int {
+    IN_X = 0,
+    IN_X_SCALE = 1,
+    IN_X_ZERO_POINT = 2,
+    IN_Y_SCALE = 3,
+    IN_Y_ZERO_POINT = 4,
+  };
+  QuantParam quant_param_;
+};
+
+// compute method of Softmax
+Status Softmax::Compute(OpKernelContext* ctx) const {
+  const auto* X = ctx->Input<Tensor>(0);
+  const auto& X_shape = X->Shape();
+  size_t rank = X_shape.NumDimensions();
+  auto* Y = ctx->Output(0, X_shape);
+
+  // edge case. one or more dims with value of 0. nothing to do
+  if (X_shape.Size() == 0) {
+    return Status::OK();
+  }
+
+  const size_t axis = static_cast<size_t>(HandleNegativeAxis(axis_, rank));
+  concurrency::ThreadPool* thread_pool = ctx->GetOperatorThreadPool();
+
+  if (opset_ < 13) {
+    return ComputeImpl(*X, *Y, axis, thread_pool);
+  } else {
+    return ComputeImplOpset13(*X, *Y, axis, thread_pool, ctx);
+  }
+}
+
+// opset-12 and below
+Status Softmax::ComputeImpl(const Tensor& input, Tensor& output, size_t axis,
+                               concurrency::ThreadPool* /*thread_pool*/) const {
+  const auto& X_shape = input.Shape();
+  const size_t N = X_shape.SizeToDimension(axis);
+  // const size_t D = X_shape.SizeFromDimension(axis);
+  xnn_status status;
+  if (kernel_dtype == ONNX_NAMESPACE::TensorProto_DataType_UINT8) {
+    status = xnn_setup_softmax_nc_qu8(
+        op0_.get(),
+        N,
+        input.template Data<uint8_t>(),
+        output.template MutableData<uint8_t>(),
+        nullptr);
+  }
+  else {
+    status = xnn_setup_softmax_nc_f32(
+        op0_.get(),
+        N,
+        input.template Data<float>(),
+        output.template MutableData<float>(),
+        nullptr);
+  }
+  ORT_ENFORCE(status == xnn_status_success, "xnn_setup_softmax_nc_type failed. Status:", status);
+  status = xnn_run_operator(op0_.get(), nullptr);
+  if (status != xnn_status_success) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "xnn_run_operator returned ", status);
+  }
+  return Status::OK();
+}
+// opset-13 and above
+Status Softmax::ComputeImplOpset13(const Tensor& input, Tensor& output, size_t axis,
+                                      concurrency::ThreadPool* thread_pool, OpKernelContext* /*ctx*/) const {
+  return ComputeImpl(input, output, axis, thread_pool);
```

Comment:
Is there some opset 13 specific logic that should be added here? Or do we not need separate branches based on the opset?

Reply:
Removed. Since softmax is not a layout-sensitive op,  It's reasonable to support only the last dim as reduced-axis.